tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -12215,7 +12215,19 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC, Expr *BitWidth,
         result = false;
       }
       break;
-    case AttributeList::AT_HLSLGroupShared:
+    case AttributeList::AT_HLSLGroupShared: {
+      const auto *shaderModel =
+        hlsl::ShaderModel::GetByName(getLangOpts().HLSLProfile.c_str());
+      // If not compute/amp/mesh, produce error
+      // If lib, give it a pass since we don't know enough to tell yet.
+      // Validation will catch it
+      if (shaderModel->IsValid() && !shaderModel->IsCS() && !shaderModel->IsAS() &&
+          !shaderModel->IsMS() && !shaderModel->IsLib()) {
+        Diag(pAttr->getLoc(), diag::err_hlsl_sm_unsupported)
+          << getLangOpts().HLSLProfile.c_str() << pAttr->getName() << pAttr->getRange();
+        result = false;
+      }
```

Comment:
I have already convinced myself that emitting an error here isn't worth it, but I want a second opinion.

The problem with emitting the error here is that any shader that has any groupshared variables at all will produce an error if it is ever built for a target that doesn't support it. That's something a number of our tests (many written by yours truly) did and it's likely that there are other such shaders out in the wild. Since we produce the error in validation only when the groupshared variable is actually used by a disallowed shader stages, it's a more accurate error that allows us to irritate people as little as possible.

Reply:
No we can't have an error here, because you are basing this on the top-level target and applying this to all code.  It's common to compile different shaders out of the same source, so this would break perfectly legal scenarios.  If we want an error before validation, it needs to be late, when we know it's actually used in the shader, after DCE.